step-api/src/main/java/org/jenkinsci/plugins/workflow/steps/AbstractSynchronousNonBlockingStepExecution.java

```diff
@@ -0,0 +1,84 @@
+package org.jenkinsci.plugins.workflow.steps;
+
+import hudson.util.DaemonThreadFactory;
+import hudson.util.NamingThreadFactory;
+
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+
+/**
+ * Similar to {@link AbstractSynchronousStepExecution} but it always executes synchronously and does not block the CPS VM thread.
+ * @see {@link StepExecution}
+ * @param <T> the type of the return value (may be {@link Void})
+ */
+public abstract class AbstractSynchronousNonBlockingStepExecution<T> extends AbstractStepExecutionImpl {
+
+    private transient volatile Future<?> task;
+
+    protected AbstractSynchronousNonBlockingStepExecution() {
+    }
+
+    protected AbstractSynchronousNonBlockingStepExecution(StepContext context) {
+        super(context);
+    }
+
+    /**
+     * Meat of the execution.
+     *
+     * When this method returns, a step execution is over.
+     */
+    protected abstract T run() throws Exception;
+
+    @Override
+    public final boolean start() throws Exception {
+        task = StepRunnerExecutorService.get().submit(new StepRunner(this));
+        return false;
+    }
+
+    /**
+     * If the computation is going synchronously, try to cancel that.
+     */
+    @Override
+    public void stop(Throwable cause) throws Exception {
+        if (task != null) {
+            task.cancel(true);
+        }
+        getContext().onFailure(cause);
```

Comment:
> Probably I should refine that so that the second abort attempt is handled in the infrastructure by calling StepContext.onFailure on the running step(s), removing the special logic in DurableTaskStep.Execution; in that case it would be safe to simply delete this line here.

So, should we wait for this?